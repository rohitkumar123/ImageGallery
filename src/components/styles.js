import {StyleSheet} from 'react-native';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    position: 'absolute',
    bottom: 0,
  },
  header: {
    marginTop: 50,
    fontSize: 30,
    textAlign: 'center',
    color: 'black',
  },
  itemStyle: {
    padding: 10,
    marginVertical: 8,
    marginHorizontal: 16,
  },
  title: {
    fontSize: 20,
    marginBottom: 10,
    textAlign: 'center',
  },
  image: {
    width: 200,
    height: 200,
    justifyContent: 'center',
    alignItems: 'center',
  },
  loadingStyle: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  error: {
    color: 'red',
  },
});

export default styles;
