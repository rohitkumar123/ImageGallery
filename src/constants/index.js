export const IMAGE_REQUEST_LOADING = 'IMAGE_REQUEST_LOADING';
export const IMAGE_REQUEST_SUCCESS = 'IMAGE_REQUEST_SUCCESS';
export const IMAGE_REQUEST_FAIL = 'IMAGE_REQUEST_FAIL';
export const RESET = 'RESET';
export const INTERNET_MESSAGE = 'Please check your internet connectivity...';
