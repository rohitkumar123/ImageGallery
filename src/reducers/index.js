import imageData from './imageReducer';

export default {
  imageData,
};
