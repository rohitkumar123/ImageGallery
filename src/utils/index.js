export const Action = (type, payload) => ({type, payload});

export const URL = (page) => {
 return `https://picsum.photos/v2/list?page=${page}&perPage=10`;
};

export const IMAGE_URL = (id) => {
  return `https://picsum.photos/200/300?image=${id}`;
};

