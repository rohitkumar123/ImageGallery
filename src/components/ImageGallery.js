import React, {useState, useEffect, useCallback} from 'react';
import {
  SafeAreaView,
  View,
  Image,
  FlatList,
  Text,
  ActivityIndicator,
} from 'react-native';
import NetInfo from '@react-native-community/netinfo';
import {useDispatch, useSelector} from 'react-redux';
import axios from 'axios';
import {Action, URL, IMAGE_URL} from '../utils';
import * as Constant from '../constants';
import styles from './styles';

const ImageGallery = () => {
  const [page, setPage] = useState(1);
  const [isNetworkError, setisNetworkError] = useState(false);
  const loading = useSelector((state) => state.imageData.loading);
  const listdata = useSelector((state) => state.imageData.data);
  const dispatch = useDispatch();

  const fetchData = useCallback(
    (page) => {
      NetInfo.fetch()
        .then((state) => {
          const {isConnected} = state;
          if (isConnected) {
            setisNetworkError(false);
            dispatch(Action(Constant.IMAGE_REQUEST_LOADING));
            axios
              .get(URL(page))
              .then((response) => {
                const {data} = response;
                dispatch(Action(Constant.IMAGE_REQUEST_SUCCESS, data));
              })
              .catch((error) => {
                dispatch(Action(Constant.IMAGE_REQUEST_FAIL, error));
              });
          } else {
            setisNetworkError(true);
          }
        })
        .catch(() => {
          setisNetworkError(true);
        });
    },
    [dispatch],
  );

  useEffect(() => {
    fetchData(page);
  }, [page]);

  const onRefresh = () => {
    setPage(page + 1);
  };

  const renderItem = ({item}) => {
    const {author, id} = item;
    return (
      <View style={styles.itemStyle}>
        <Text style={styles.title}>{author}</Text>
        <Image
          style={styles.image}
          source={{
            uri: IMAGE_URL(id),
          }}
        />
      </View>
    );
  };
  const Loading = () => {
    return (
      <>
        {loading && (
          <View style={styles.loadingStyle}>
            <ActivityIndicator size="large" color="#0000ff" />
          </View>
        )}
      </>
    );
  };
  return (
    <>
      <Text style={styles.header}>Image Gallery</Text>
      {isNetworkError && (
        <Text style={[styles.header, styles.error]}>
          {Constant.INTERNET_MESSAGE}
        </Text>
      )}
      {!isNetworkError && Loading()}
      <SafeAreaView style={styles.container}>
        <FlatList
          horizontal
          data={listdata}
          renderItem={renderItem}
          keyExtractor={(item, index) => `key-${item.id}`}
          onRefresh={() => onRefresh()}
          refreshing={loading}
        />
      </SafeAreaView>
    </>
  );
};

export default ImageGallery;
