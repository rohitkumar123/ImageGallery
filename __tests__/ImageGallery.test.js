import React from 'react';
import {ActivityIndicator, FlatList, View, Text, Image} from 'react-native';
import configureMockStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import thunk from 'redux-thunk';
import renderer from 'react-test-renderer';
import ImageGallery from '../src/components/ImageGallery';
import styles from '../src/components/styles';
import {IMAGE_URL} from '../src/utils';

jest.mock('axios');
jest.mock('react-redux');
jest.mock('@react-native-community/netinfo', () => ({
  fetch: () => Promise.resolve(jest.fn()),
}));

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('My Connected React-Redux Component', () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({
      myState: 'sample text',
    });

    component = renderer.create(
      <Provider store={store}>
        <ImageGallery />
      </Provider>,
    );
  });

  it('should render with given state from Redux store', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});
it('renders correctly', () => {
  const tree = renderer.create(<ImageGallery />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders the ActivityIndicator component', () => {
  const tree = renderer
    .create(<ActivityIndicator animating={true} size="large" color="#0000ff" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders the FlatList component', () => {
  const tree = renderer
    .create(
      <FlatList
        refreshing={true}
        onRefresh={() => {}}
        data={[
          {
            format: 'jpeg',
            width: 5616,
            height: 3744,
            filename: '0.jpeg',
            id: 0,
            author: 'Alejandro Escamilla',
            author_url: 'https://unsplash.com/photos/yC-Yzbqy7PY',
            post_url: 'https://unsplash.com/photos/yC-Yzbqy7PY',
          },
          {
            format: 'jpeg',
            width: 5616,
            height: 3744,
            filename: '1.jpeg',
            id: 1,
            author: 'Alejandro Escamilla',
            author_url: 'https://unsplash.com/photos/LNRyGwIJr5c',
            post_url: 'https://unsplash.com/photos/LNRyGwIJr5c',
          },
        ]}
        keyExtractor={(item) => `key-${item.id}`}
        renderItem={({item}) => {
          const {author, id} = item;
          return (
            <View style={styles.item}>
              <Text style={styles.title}>{author}</Text>
              <Image
                style={styles.image}
                source={{
                  uri: IMAGE_URL(id),
                }}
              />
            </View>
          );
        }}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
