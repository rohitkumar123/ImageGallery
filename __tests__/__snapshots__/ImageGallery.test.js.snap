// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <Text
    style={
      Object {
        "color": "black",
        "fontSize": 30,
        "marginTop": 50,
        "textAlign": "center",
      }
    }
  >
    Image Gallery
  </Text>,
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "flex": 1,
        "justifyContent": "center",
        "position": "absolute",
      }
    }
  >
    <RCTScrollView
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={true}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onRefresh={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <RefreshControlMock
          onRefresh={[Function]}
          refreshing={false}
        />
      }
      refreshing={false}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <RCTRefreshControl />
      <View />
    </RCTScrollView>
  </RCTSafeAreaView>,
]
`;

exports[`renders the ActivityIndicator component 1`] = `
<ActivityIndicator
  animating={true}
  color="#0000ff"
  hidesWhenStopped={true}
  size="large"
/>
`;

exports[`renders the FlatList component 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "author": "Alejandro Escamilla",
        "author_url": "https://unsplash.com/photos/yC-Yzbqy7PY",
        "filename": "0.jpeg",
        "format": "jpeg",
        "height": 3744,
        "id": 0,
        "post_url": "https://unsplash.com/photos/yC-Yzbqy7PY",
        "width": 5616,
      },
      Object {
        "author": "Alejandro Escamilla",
        "author_url": "https://unsplash.com/photos/LNRyGwIJr5c",
        "filename": "1.jpeg",
        "format": "jpeg",
        "height": 3744,
        "id": 1,
        "post_url": "https://unsplash.com/photos/LNRyGwIJr5c",
        "width": 5616,
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onRefresh={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={true}
    />
  }
  refreshing={true}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "marginHorizontal": 16,
            "marginVertical": 8,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
              "marginBottom": 10,
              "textAlign": "center",
            }
          }
        >
          Alejandro Escamilla
        </Text>
        <Image
          source={
            Object {
              "uri": "https://picsum.photos/200/300?image=0",
            }
          }
          style={
            Object {
              "alignItems": "center",
              "height": 200,
              "justifyContent": "center",
              "width": 200,
            }
          }
        />
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "marginHorizontal": 16,
            "marginVertical": 8,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
              "marginBottom": 10,
              "textAlign": "center",
            }
          }
        >
          Alejandro Escamilla
        </Text>
        <Image
          source={
            Object {
              "uri": "https://picsum.photos/200/300?image=1",
            }
          }
          style={
            Object {
              "alignItems": "center",
              "height": 200,
              "justifyContent": "center",
              "width": 200,
            }
          }
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;
